crypto/evp/e_aes.c

```diff
@@ -796,7 +816,30 @@ static int aes_t4_xts_init_key(EVP_CIPHER_CTX *ctx, const unsigned char *key,
         return 1;
 
     if (key) {
-        int bits = EVP_CIPHER_CTX_key_length(ctx) * 4;
+        /* The key is two half length keys in reality */
+        const int bits = EVP_CIPHER_CTX_key_length(ctx) * 8 / 2;
+
+        /*
+         * Verify that the two keys are different.
+         *
+         * This addresses the vulnerability described in Rogaway's
+         * September 2004 paper:
+         *
+         *      "Efficient Instantiations of Tweakable Blockciphers and
+         *       Refinements to Modes OCB and PMAC".
+         *      (http://web.cs.ucdavis.edu/~rogaway/papers/offsets.pdf)
+         *
+         * FIPS 140-2 IG A.9 XTS-AES Key Generation Requirements states
+         * that:
+         *      "The check for Key_1 != Key_2 shall be done at any place
+         *       BEFORE using the keys in the XTS-AES algorithm to process
+         *       data with them."
+         */
```

Comment:
Three times the same long comment seems a bit excessive. Wouldn't it be better to only have it once, and subsequently refer to it?

```C
        /*
         * Verify that the two keys are different.
         * 
         * This addresses Rogaway's vulnerability, see previous comment.
         */
```

Reply:
Done.